### DISCUSSION TITLE
Data importer: creating a custom mapping operator

### DISCUSSION THREAD

**Discussion by sissonen** [pimcore/pimcore#9781] [Q&A / Support]
Hi!

I am migrating Pimcore 6.8 CSV imports to Data importer data sources and I need some support to get a custom mapping operator to work. Previously I used "PHP code" operator but that is no longer available. Here is what I have done so far:

1. I made a ConditionalTextConversion.php file based on StaticText.php in data-importer. The file I placed in src/Datahub/Operators with namespace App\Datahub\Operators\ConditionalTextConversion.
2. To config/services.yaml I added:
`App\Datahub\Operators\ConditionalTextConversion:
        tags:
            - { name: "pimcore.datahub.data-importer.operator", type: "conditionalTextConversion" }`
3. I made conditionalTextConversion.js based on staticText.js in public/bundles/pimcoredataimporter/js/pimcore/configuration/components/mapping/operator

When running "bin/console debug:container --show-hidden conditionalTextConversion" I get similar information on the service as is returned for staticText -service.

But the new operator does not appear in the operators list in the mapping operator configuration. I feel like there is something missing that I should do but I cannot find information what it could be. Could someone help out?

**Comment by fashxp** [comment]
is the JS file loaded? 

see for example [here](https://pimcore.com/docs/pimcore/current/Development_Documentation/Extending_Pimcore/Bundle_Developers_Guide/Loading_Admin_UI_Assets.html) for details

**Reply by sissonen** [reply]
The JS was not loaded.
I created an EventListener based on the class on the page you linked and added a service:
```
    App\EventListener\AdminAssetsListener:
        tags:
            - { name: kernel.event_listener, event: pimcore.bundle_manager.paths.js, method: onJsPaths }

```
The new operator still does not appear and Symfony profiler has a log entry under "Container":
Symfony\Component\DependencyInjection\Compiler\AutowirePass
Skipping service "App\Datahub\Operators\ConditionalTextConversion": Class or interface "App\Datahub\Operators\ConditionalTextConversion" cannot be loaded.

Any idea what could still be wrong?

**Reply by fashxp** [reply]
What is the classname of the js file? Needs to be in same namespace as the ones of the bundle...

**Reply by sissonen** [reply]
I just now noticed that the directories under public/bundles/ are symlinks to vendor directories. So, I suppose a custom operator JS-file should not be in public/bundles/pimcoredataimporter/js/pimcore/configuration/components/mapping/operator ?

The JS file contents (based on staticText.js):

```
pimcore.registerNS("pimcore.plugin.pimcoreDataImporterBundle.configuration.components.mapping.operator.conditionalTextConversion");
pimcore.plugin.pimcoreDataImporterBundle.configuration.components.mapping.operator.conditionalTextConversion = Class.create(pimcore.plugin.pimcoreDataImporterBundle.configuration.components.mapping.abstractOperator, {

    type: 'conditionalTextConversion',

    getFormItems: function() {
        return [
            {
                xtype: 'textfield',
                fieldLabel: t('plugin_pimcore_datahub_data_importer_configpanel_transformation_pipeline_text'),
                value: this.data.settings ? this.data.settings.original
                name: 'settings.original',
                listeners: {
                    change: this.inputChangePreviewUpdate.bind(this)
                }

            },

            {
                xtype: 'textfield',
                fieldLabel: t('plugin_pimcore_datahub_data_importer_configpanel_transformation_pipeline_text'),
                value: this.data.settings ? this.data.settings.converted : '',
                name: 'settings.converted',
                listeners: {
                    change: this.inputChangePreviewUpdate.bind(this)
                }
            }
        ];
    }

});
```

**Reply by fashxp** [reply]
The file looks good, but cannot be in `public/bundles/pimcoredataimporter/js/pimcore/configuration/components/mapping/operator ` ... as this comes from the bundle it self. 
Your custom file needs to be located in the public directory of your bundle or somewhere in the global public directory if not in a bundle. 

BR
Christian

**Reply by sissonen** [reply]
I moved the js out of the vendor dir to global public dir, and fixed other things until I got the new operator to be shown in the list. But then I bumped into "ERROR: Unknown operator type `conditionalTextConversion`" in the "Transformation Result" panel. This has probably something to do with the namespaces but could not figure what the problem was exactly.

So, I decided to fix this by making the new operator into the bundle (I only need the import to run a few times so losing the changes in a future update is not so bad) and got it working quite easily. If you would be interested in consolidating the operator into the data-importer bundle I can provide the code (it will probably need some more work on it, though).

I also went with a similar solution with "Load DataObject" operator where I need to match a existing Object with a partial match to an attribute value (use case: product category mapping where imported products have a different category system from the one configured in Pimcore -> Pimcore categories can have multiple imported categories mapped to them). I can also share this solution if you are interested.

Thanks for all your help Christian! :)

**Reply by fashxp** [reply]
to be honest, I'm not sure, if I understood what your operators exactly do. 
but if you have already the code in place and are ready to share - just create a PR, then we continue the discussion there. we are always open for and happy about good contributions 😊

**Reply by sissonen** [reply]
I created a PR (my first ever :sweat_smile: ) for the conditional conversion operator. Is it OK?

**Reply by fashxp** [reply]
hopefully it is not the last one :)
